harness/determined/profiler.py

```diff
@@ -0,0 +1,559 @@
+        return timestamp.isoformat() + "Z"
+
+    def convert_to_post_format(self) -> List["TrialProfilerMetricsBatch"]:
+        def to_post_format(
+            measurements: List[Any], labels: Dict[str, Any]
+        ) -> "TrialProfilerMetricsBatch":
+            values, batches, timestamps = [], [], []
+            for m in measurements:
+                values.append(m.measurement)
+                batches.append(m.batch_index)
+                timestamps.append(self.convert_to_timestamp_str(m.timestamp))
+            return TrialProfilerMetricsBatch(values, batches, timestamps, labels)
+
+        def make_labels(name: str, metric_type: str, gpu_uuid_label: str = "") -> Dict[str, Any]:
+            return {
+                "trialId": self.trial_id,
+                "name": name,
+                "agentId": self.agent_id,
+                "gpuUuid": gpu_uuid_label,
+                "metricType": metric_type,
+            }
+
+        trial_profiler_metrics_batches = []
+        for metric_name in self.batch.keys():
+            # TODO: Don't forget to include GPU Memory
+            if metric_name != SysMetricType.GPU_UTIL_METRIC and len(self.batch[metric_name]) > 0:
+                trial_profiler_metrics_batches.append(
+                    to_post_format(
+                        self.batch[metric_name],
+                        make_labels(metric_name, SYSTEM_METRIC_TYPE_ENUM),
+                    )
+                )
+
+            # GPU Metrics need to be grouped by GPU UUID
+            # TODO: Don't forget to include GPU Memory
+            if (
+                metric_name == SysMetricType.GPU_UTIL_METRIC
+                and len(self.batch[metric_name].keys()) > 0
+            ):
+                for gpu_uuid in self.batch[metric_name].keys():
+                    trial_profiler_metrics_batches.append(
+                        to_post_format(
+                            self.batch[metric_name][gpu_uuid],
+                            make_labels(
+                                metric_name, SYSTEM_METRIC_TYPE_ENUM, gpu_uuid_label=gpu_uuid
+                            ),
+                        )
+                    )
+
+        return trial_profiler_metrics_batches
+
+
+class StartMessage:
+    pass
+
+
+class ShutdownMessage:
+    pass
+
+
+class SysMetricCollectorThread(threading.Thread):
+    """
+    Background thread for collecting profiler metrics at a high granularity and shipping them to
+    the master
+
+    - SimpleCpuUtilization = Measured in percent
+    - FreeMemory = Measured in Gigabytes
+    - NetworkSentThroughput = Measured in Gigabit/s
+    - NetworkRecvThroughput = Measured in Gigabit/s
+    - DiskIops
+    - DiskReadThroughput = Measured in bytes/second
+    - DiskWriteThroughput = Measured in bytes/second
+    - GpuUtilization = Measured in percent
+    """
+
+    FLUSH_INTERVAL = 10  # How often to make API calls
+    MEASUREMENT_INTERVAL = 0.1
```

Comment:
q: should this be configurable?

Reply:
It should be configurable by us in the future, but it doesn't not need to be configurable by users